File: frontend/src/api/risk.ts
import apiClient from './client';
import type { ApiResponse } from './types';

export interface RiskStatus {
  level: string;
  isLocked: boolean;
  lockReason: string | null;
  currentDrawdown: number;
  dailyLoss: number;
  consecutiveLosses: number;
  maxDrawdown: number;
  maxDailyLoss: number;
  maxConsecutiveLosses: number;
}

export interface RiskEvent {
  eventId: string;
  timestamp: string;
  eventType: string;
  severity: 'info' | 'warning' | 'critical';
  message: string;
  details?: Record<string, unknown>;
}

export interface RiskEventsResponse {
  events: RiskEvent[];
  total: number;
}

// 后端返回格式
interface BackendRiskStatus {
  level: string;
  is_locked: boolean;
  lock_reason: string | null;
  lock_since?: string;
  recent_events: any[];
  daily_loss: number;
  current_drawdown: number;
}

interface BackendRiskEvent {
  event_id: string;
  timestamp: string;
  event_type: string;
  severity: string;
  message: string;
  details?: Record<string, unknown>;
}

export const riskApi = {
  getStatus: async (): Promise<RiskStatus> => {
    try {
      const res = await apiClient.get<ApiResponse<BackendRiskStatus>>('/api/v1/risk/status');
      if (!res.data.success || !res.data.data) {
        return {
          level: 'NORMAL',
          isLocked: false,
          lockReason: null,
          currentDrawdown: 0,
          dailyLoss: 0,
          consecutiveLosses: 0,
          maxDrawdown: 20,
          maxDailyLoss: 3,
          maxConsecutiveLosses: 3,
        };
      }
      const d = res.data.data;
      return {
        level: d.level,
        isLocked: d.is_locked,
        lockReason: d.lock_reason,
        currentDrawdown: d.current_drawdown || 0,
        dailyLoss: d.daily_loss || 0,
        consecutiveLosses: 0,
        maxDrawdown: 20,
        maxDailyLoss: 3,
        maxConsecutiveLosses: 3,
      };
    } catch {
      return {
        level: 'NORMAL',
        isLocked: false,
        lockReason: null,
        currentDrawdown: 0,
        dailyLoss: 0,
        consecutiveLosses: 0,
        maxDrawdown: 20,
        maxDailyLoss: 3,
        maxConsecutiveLosses: 3,
      };
    }
  },

  getEvents: async (limit = 50): Promise<RiskEventsResponse> => {
    try {
      const res = await apiClient.get<ApiResponse<BackendRiskEvent[]>>('/api/v1/risk/events', {
        params: { limit },
      });
      if (!res.data.success || !res.data.data) {
        return { events: [], total: 0 };
      }
      const events = res.data.data.map((e) => ({
        eventId: e.event_id,
        timestamp: e.timestamp,
        eventType: e.event_type,
        severity: (e.severity === 'CRITICAL' ? 'critical' : e.severity === 'WARNING' ? 'warning' : 'info') as RiskEvent['severity'],
        message: e.message,
        details: e.details,
      }));
      return { events, total: events.length };
    } catch {
      return { events: [], total: 0 };
    }
  },
};

File: backend/src/api/routes/risk.py
"""
BTC 自动交易系统 — 风控路由

提供风控状态查询和管理接口。
"""

from datetime import timedelta
from uuid import uuid4

from fastapi import APIRouter, HTTPException, status

from src.api.auth import CurrentUser, CurrentAdmin, audit_log
from src.api.dependencies import RiskEngineDep
from src.api.schemas import (
    ApiResponse,
    RiskEventResponse,
    RiskStatusResponse,
    UnlockRequest,
)
from src.common.enums import RiskLevel
from src.common.utils import utc_now

router = APIRouter(prefix="/risk", tags=["风控"])


@router.get("/status", response_model=ApiResponse[RiskStatusResponse])
async def get_risk_status(
    user: CurrentUser,
    risk_engine: RiskEngineDep,
) -> ApiResponse[RiskStatusResponse]:
    """
    获取风控状态
    
    返回当前风控级别、锁定状态和最近事件。
    """
    # 获取风控状态
    is_locked = risk_engine.is_locked
    lock_reason = risk_engine.lock_reason if is_locked else None
    
    # 简化：实际应从风控引擎获取
    recent_events = []
    
    response = RiskStatusResponse(
        level=RiskLevel.NORMAL if not is_locked else RiskLevel.CRITICAL,
        is_locked=is_locked,
        lock_reason=lock_reason,
        lock_since=utc_now() if is_locked else None,
        recent_events=recent_events,
        daily_loss=0.0,
        current_drawdown=0.0,
    )
    
    return ApiResponse(data=response)


@router.get("/events", response_model=ApiResponse[list[RiskEventResponse]])
async def get_risk_events(
    user: CurrentUser,
    risk_engine: RiskEngineDep,
    limit: int = 50,
    severity: RiskLevel | None = None,
) -> ApiResponse[list[RiskEventResponse]]:
    """
    获取风控事件
    
    返回最近的风控事件列表。
    """
    # 简化实现：实际应从存储获取
    events = [
        RiskEventResponse(
            event_id=str(uuid4()),
            timestamp=utc_now() - timedelta(hours=1),
            event_type="daily_loss_warning",
            severity=RiskLevel.WARNING,
            message="日亏损接近阈值",
            details={"current_loss": 0.025, "threshold": 0.03},
        ),
    ]
    
    # 按严重程度过滤
    if severity:
        events = [e for e in events if e.severity == severity]
    
    return ApiResponse(data=events[:limit])


@router.post("/unlock", response_model=ApiResponse[RiskStatusResponse])
async def request_unlock(
    admin: CurrentAdmin,
    risk_engine: RiskEngineDep,
    request: UnlockRequest,
) -> ApiResponse[RiskStatusResponse]:
    """
    请求解锁
    
    需要管理员权限。尝试解除风控锁定。
    """
    audit_log.log(
        api_key=admin.user_id,
        action="request_unlock",
        resource="risk_engine",
        details={"reason": request.reason},
    )
    
    if not risk_engine.is_locked:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail={"code": "NOT_LOCKED", "message": "系统未被锁定"},
        )
    
    # 尝试解锁
    try:
        risk_engine.unlock(request.reason)
        
        response = RiskStatusResponse(
            level=RiskLevel.NORMAL,
            is_locked=False,
            lock_reason=None,
            lock_since=None,
            recent_events=[],
            daily_loss=0.0,
            current_drawdown=0.0,
        )
        
        return ApiResponse(data=response)
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_403_FORBIDDEN,
            detail={"code": "UNLOCK_DENIED", "message": str(e)},
        )
